lib/components/ApproveReject.js

```diff
@@ -3,70 +3,140 @@ import PropTypes from 'prop-types';
 import { Text, Input, Button, Dropdown } from '@bpanel/bpanel-ui';
 import BoxGrid from './BoxGrid';
 import Label from './Label';
+import { HARDWARE_TYPES } from '../constants';
 
 class ApproveReject extends PureComponent {
   constructor() {
     super();
     this.state = {
       cosignerToken: '',
-      selectedProposal: {},
+      selectedHardware: '',
+      accountNumber: '',
+      proposal: {},
     };
   }
 
   static get propTypes() {
     return {
       selectedWallet: PropTypes.string,
       selectedProposal: PropTypes.object,
+      approveProposal: PropTypes.func,
+      rejectProposal: PropTypes.func,
+      proposalDropdown: PropTypes.array,
     };
   }
 
   static get defaultProps() {
-    return {};
+    return {
+      proposalDropdown: [],
+    };
   }
 
   update(key, value) {
     this.setState({ [key]: value });
   }
 
   approve() {
-    console.log('Not implemented!');
+    const { approveProposal, selectedWallet } = this.props;
+    const {
+      accountNumber,
+      cosignerToken,
+      selectedHardware,
+      proposal,
+    } = this.state;
+
+    if (selectedHardware === '') {
+      alert('please select hardware wallet type');
+      return;
+    }
+    if (accountNumber === '') {
+      alert('please enter account number');
+      return;
+    }
+
+    approveProposal(selectedWallet, proposal.name, {
+      hardwareType: selectedHardware,
+      cosignerToken,
+      account: accountNumber,
+    });
   }
 
   reject() {
-    console.log('Not implemented!');
+    const { rejectProposal, selectedWallet } = this.props;
+    const { proposal, cosignerToken } = this.state;
+    rejectProposal(selectedWallet, proposal.name, cosignerToken);
+  }
+
+  renderProposalInfo() {
+    const { proposal } = this.state;
```

Comment:
This also should display detailed proposal info, such as the tx amount, recipient

Reply:
Yup, this is going to another PR so that this one doesn't grow too large